lib/srv/regular/sshserver.go

```diff
@@ -1096,40 +1109,254 @@ func (s *Server) getServerResource() (types.Resource, error) {
+		err = f(fd)
+	})
+	return trace.NewAggregate(err2, err)
+}
+
+// validateListenerSocket checks that the socket and listener file descriptor
+// sent from the forwarding process have the expected properties.
+func validateListenerSocket(scx *srv.ServerContext, controlConn *net.UnixConn, listenerFD *os.File) error {
+	// Get the credentials of the client connected to the socket.
+	syscallConn, err := controlConn.SyscallConn()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var cred *unix.Ucred
+	if err := controlSyscallConn(syscallConn, func(fd uintptr) error {
+		cred, err = unix.GetsockoptUcred(int(fd), unix.SOL_SOCKET, unix.SO_PEERCRED)
+		return err
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Check that the user connected to the socket is who we expect.
+	usr, err := user.Lookup(scx.Identity.Login)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if expectedUid, err := strconv.Atoi(usr.Uid); err != nil {
+		return trace.Wrap(err)
+	} else if int(cred.Uid) != expectedUid {
+		return trace.AccessDenied("unexpected user UID for the socket: %v", cred.Uid)
+	}
+	if expectedGid, err := strconv.Atoi(usr.Gid); err != nil {
+		return trace.Wrap(err)
+	} else if int(cred.Gid) != expectedGid {
+		return trace.AccessDenied("unexpected user GID for the socket: %v", cred.Gid)
+	}
+
+	// Verify that the listener is a socket.
+	if fileInfo, err := listenerFD.Stat(); err != nil {
+		return trace.Wrap(err)
+	} else if fileInfo.Mode()&os.ModeSocket == 0 {
+		return trace.AccessDenied("file descriptor is not a socket")
+	}
+
+	// Verify the socket type
+	if sockType, err := unix.GetsockoptInt(int(listenerFD.Fd()), unix.SOL_SOCKET, unix.SO_TYPE); err != nil {
+		return trace.Wrap(err)
+	} else if sockType != unix.SOCK_STREAM && sockType != unix.SOCK_SEQPACKET {
+		return trace.AccessDenied("socket is not of the expected type (STREAM or SEQPACKET)")
+	}
+
+	// Verify that reuse is not enabled on the socket
+	if reuseAddr, err := unix.GetsockoptInt(int(listenerFD.Fd()), unix.SOL_SOCKET, unix.SO_REUSEADDR); err != nil {
+		return trace.Wrap(err)
+	} else if reuseAddr != 0 {
+		return trace.AccessDenied("SO_REUSEADDR is enabled on the socket")
+	}
+	if reusePort, err := unix.GetsockoptInt(int(listenerFD.Fd()), unix.SOL_SOCKET, unix.SO_REUSEPORT); err != nil {
+		// Some systems may not support SO_REUSEPORT, so we ignore the error here
+	} else if reusePort != 0 {
+		return trace.AccessDenied("SO_REUSEPORT is enabled on the socket")
+	}
+
+	// Verify file descriptor flags
+	if fdFlags, err := unix.FcntlInt(listenerFD.Fd(), unix.F_GETFL, 0); err != nil {
+		return trace.Wrap(err)
+	} else if fdFlags&unix.O_NONBLOCK != 0 {
+		return trace.AccessDenied("socket is set to non-blocking mode")
+	}
+	if fdFlags, _, err := unix.Syscall(unix.SYS_FCNTL, listenerFD.Fd(), unix.F_GETFD, 0); err != 0 {
+		return trace.Wrap(err)
+	} else if fdFlags&unix.FD_CLOEXEC == 0 {
+		return trace.AccessDenied("socket is not set to close on exec")
+	}
```

Comment:
This is not something that we need to check, the file descriptor is private to the current process, the utility functions in `uds` already make sure that the file descriptor is spawned as cloexec as it's read through the socket, and `net.FileListener` is guaranteed to duplicate it cloexec.